src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -783,6 +637,78 @@ private Compilation CreateEmptyCompilation()
                 return finalState.HasSuccessfullyLoaded;
             }
 
+            public ICompilationTracker FreezePartialState(CancellationToken cancellationToken)
+            {
+                var state = this.ReadState();
+
+                // If we're finalized and already frozen, we can just use ourselves.
+                if (state is FinalCompilationTrackerState { IsFrozen: true } finalState)
+                    return this;
+
+                var projectState = state switch
+                {
+                    // If we don't have an existing state, then transition to a project state without any data.
+                    null => this.ProjectState.RemoveAllDocuments(),
+                    FinalCompilationTrackerState => this.ProjectState,
+                    // If we have an in progress state with no steps, then we're just at the current project state.
```

Comment:
Nit: I wish there were blank lines here before the comments to split up the different cases more, but I'm willing to assume this looks much bettter than on GitHub where the comments are basically the same color as text.

Reply:
done.